worlds/sc2wol/PoolFilter.py

```diff
@@ -132,6 +132,64 @@ def generate_reduced_inventory(self, inventory_size: int, mission_requirements:
         cascade_keys = self.cascade_removal_map.keys()
         units_always_have_upgrades = get_option_value(self.multiworld, self.player, "units_always_have_upgrades")
 
+        def attempt_removal(item: Item) -> bool:
+            # If item can be removed and has associated items, remove them as well
+            inventory.remove(item)
+            # Only run logic checks when removing logic items
+            if item.name in self.logical_inventory:
+                self.logical_inventory.remove(item.name)
+                if not all(requirement(self) for requirement in requirements):
+                    # If item cannot be removed, lock or revert
+                    self.logical_inventory.add(item.name)
+                    for _ in range(get_item_quantity(item)):
+                        locked_items.append(copy_item(item))
+                    return False
+            return True
+        
+        # Limit the maximum number of upgrades 
+        maxUpgrad = get_option_value(self.multiworld, self.player, 
+                            "max_nb_upgrades")
+        if maxUpgrad != -1:
+            unit_avail_upgrades = {}
+            # Needed to take into account locked/existing items
+            unit_nb_upgrades = {}
+            for item in inventory:
+                cItem = get_full_item_list()[item.name]
+                if cItem.type == "Unit" and item.name not in unit_avail_upgrades:
+                    unit_avail_upgrades[item.name] = []
+                    unit_nb_upgrades[item.name] = 0
+                elif cItem.parent_item is not None:
+                    if cItem.parent_item not in unit_avail_upgrades:
+                        unit_avail_upgrades[cItem.parent_item] = [item]
+                        unit_nb_upgrades[cItem.parent_item] = 1
+                    else:
+                        unit_avail_upgrades[cItem.parent_item].append(item)
+                        unit_nb_upgrades[cItem.parent_item] += 1
+            # For those two categories, we count them but dont include them in removal
+            for item in locked_items + self.existing_items:
+                cItem = get_full_item_list()[item.name]
+                if cItem.type == "Unit" and item.name not in unit_avail_upgrades:
+                    unit_avail_upgrades[item.name] = []
+                    unit_nb_upgrades[item.name] = 0
+                elif cItem.parent_item is not None:
+                    if cItem.parent_item not in unit_avail_upgrades:
+                        unit_nb_upgrades[cItem.parent_item] = 1
+                    else:
+                        unit_nb_upgrades[cItem.parent_item] += 1
+            # Making sure that the upgrades being removed is random 
+            # Currently, only for combat shield vs Stabilizer Medpacks...
+            suffled_unit_list = list(unit_avail_upgrades.keys())
+            self.multiworld.random.shuffle(suffled_unit_list)
+            for unit in suffled_unit_list:
```

Comment:
shuffled_unit_upgrade_list. Not list of units but a list of unit upgrades

Reply:
Applied!